Adaptors/Amqp/src/SessionAmqp.cs

```diff
@@ -0,0 +1,161 @@
+    }
+  }
+
+  public async Task<SessionAmqp> OpenConnection()
+  {
+    if (Session.SessionState != SessionState.Opened)
+    {
+      await Session.CloseAsync()
+                   .ConfigureAwait(false);
+      if (retriesReconnect_ <= 0)
+      {
+        throw new AmqpException(new Error($"After {Options.MaxRetry} retries. Fail to reconnect to Amqp"));
+      }
+
+      --retriesReconnect_;
+    }
+
+    var address = new Address(Options.Host,
+                              Options.Port,
+                              Options.User,
+                              Options.Password,
+                              scheme: Options.Scheme);
+
+    var connectionFactory = new ConnectionFactory();
+    if (Options.Scheme.Equals("AMQPS"))
+    {
+      connectionFactory.SSL.RemoteCertificateValidationCallback = delegate(object          sender,
+                                                                           X509Certificate certificate,
+                                                                           X509Chain       chain,
+                                                                           SslPolicyErrors errors)
+                                                                  {
+                                                                    switch (errors)
+                                                                    {
+                                                                      case SslPolicyErrors.RemoteCertificateNameMismatch when Options.AllowHostMismatch:
+                                                                      case SslPolicyErrors.None:
+                                                                        return true;
+                                                                      default:
+                                                                        Logger.LogError("SSL error : {error}",
+                                                                                        errors);
+                                                                        return false;
+                                                                    }
+                                                                  };
+    }
+
+    var retry = 0;
+    for(; retry < Options.MaxRetry; retry++)
+    {
+      try
+      {
+        var connection = await connectionFactory.CreateAsync(address)
+                                                .ConfigureAwait(false);
+        connection.AddClosedCallback((x,
+                                      e) => OnCloseConnection(x,
+                                                              e,
+                                                              Logger));
+        Session = new Session(connection);
+        break;
+      }
+      catch
+      {
+        Thread.Sleep(1000 * retry);
+      }
+    }
+
+    if (retry == Options.MaxRetry)
+    {
+      throw new ArgumentOutOfRangeException(nameof(Options.MaxRetry),
```

Comment:
I think `Timeout` is a better name. I don't know if there already is a dotnet exception like that, but I guess so.